### PR TITLE
client: instance types

### PR DESCRIPTION
**Pull request by tomponline**
This PR continues what was started in https://github.com/lxc/lxd/pull/6193 

- Adds a full set of instance management functions to the `InstanceServer` interface.
- Adds the implementations of those functions that use the `/1.0/instances` endpoints if supported by the server or falls back to using `/1.0/containers`.

### CODE REVIEW

**Comment by tomponline** [comment]
@stgraber I've rebased this on your PR https://github.com/lxc/lxd/pull/6196

**Comment by stgraber** [comment]
We don't want any of those new functions to rely on the structs from interfaces_legacy, so we need updated copies of all the arg structs for instances

**Comment by tomponline** [comment]
@stgraber ah ok, misunderstood, I removed them this morning. So its basically the same as it was yesterday without the functions that accepted or returned the ContainerServer.

**Comment by stgraber** [comment]
Sorry for the confusion :)

The plan is that with LXD 4.0 we should be able to do:
 - Remove lxd_containers.go
 - Remove instances_legacy.go
 - Drop the `*Container*` functions from interfaces.go

And with no additional changes we should find ourselves with a perfectly working Go client without any of the legacy stuff. We do need to send a branch which will appropriately mark all of those structs and functions as deprecated though and mention that they will go away with 4.0 so that readers of our godoc documentation don't get tempted to use any of them :)

**Comment by tomponline** [comment]
@stgraber that has been reworked now as discussed.

File: client/interfaces.go [hunk 1]
```diff
@@ -136,6 +136,63 @@ type InstanceServer interface {
 	UpdateContainerTemplateFile(containerName string, templateName string, content io.ReadSeeker) (err error)
 	DeleteContainerTemplateFile(name string, templateName string) (err error)
 
+	// Instance functions.
+	GetInstanceNames(instanceType api.InstanceType) (names []string, err error)
+	GetInstances(instanceType api.InstanceType) (instances []api.Instance, err error)
+	GetInstancesFull(instanceType api.InstanceType) (instances []api.InstanceFull, err error)
+	GetInstance(name string) (instance *api.Instance, ETag string, err error)
+	CreateInstance(instance api.InstancesPost) (op Operation, err error)
+	CreateInstanceFromImage(source ImageServer, image api.Image, imgcontainer api.InstancesPost) (op RemoteOperation, err error)
```
Comment (stgraber) [inline client/interfaces.go]: Should change `imgContainer` there for `req` I think

Comment (tomponline) [reply client/interfaces.go]: @stgraber will do

Comment (tomponline) [reply client/interfaces.go]: @stgraber done

**Pull request merged**